### DISCUSSION TITLE
Proposal for dropping ie11 support in Vue 3

### DISCUSSION THREAD

**Discussion by yyx990803** [vuejs/rfcs#296] [RFC Discussions]
## Status

**This RFC has been finalized and merged.**

## Summary

- Drop IE11 support plan for Vue 3.
- Focus on backport compatible features back to Vue 2.7.

## Links

- [Full Rendered Proposal](https://github.com/vuejs/rfcs/blob/ie11/active-rfcs/0000-vue3-ie11-support.md)
- [RFC PR](https://github.com/vuejs/rfcs/pull/294)

**Comment by CyberAP** [comment]
I think an estimate on 2.7 release date would be very helpful for those who plan to support IE.

**Reply by yyx990803** [reply]
No estimates can be made at this point because the scope of features to be backported are not finalized yet. If only targeting a minimal feature set aimed at bringing Composition API into Vue 2 core, then it will probably be around end of Q3 2021.

**Reply by CyberAP** [reply]
How can community help with the 2.7 release to make it happen sooner?

**Reply by yyx990803** [reply]
There are some initial steps that we need to figure out (i.e. the best way to merge `@vue/composition-api` into the Vue 2 codebase), but once we setup the initial port/test infrastructure you are welcome to contribute!

**Comment by ghost** [comment]
![no-internet-explorer](https://user-images.githubusercontent.com/11320080/113430339-abd9c600-93e2-11eb-9f66-bceb323d25e0.png)

**Reply by lxyAK** [reply]
Meet the future reject IE

**Reply by adivolclone** [reply]
it's true to drop IE!

**Comment by groundhogs** [comment]
What is gained from it exactly on on a framework level?

**Reply by seanogdev** [reply]
Not having to pollute their codebase for a browser thats on it's last legs? And a browser which doesnt support the Vue 3 reactivity system out of the box anyway

**Reply by Akryum** [reply]
We wrote a Motivations section [here](https://github.com/vuejs/rfcs/blob/ie11/active-rfcs/0000-vue3-ie11-support.md#motivation)

**Reply by yyx990803** [reply]
Please read the full RFC.

**Comment by seanogdev** [comment]
I think this is a great idea.

We're stuck on Vue 2 not because of IE11, but because of waiting on the ecosystem to catch-up. As a stop-gap we've been using `@vue/composition-api` and `vue-demi` to help scaffold out a new application, with a view to migrating to Vue 3 in the short to medium term.

I have two questions:

- Do you consider the `@vue/composition-api` to be feature-complete or are there more apis that need/could be backported in vue 2.7?
- Will upgrading to 2.7 make it easier to migrate to Vue 3? Will there be an investment in a codemod migrating from the composables written using vue 2 to vue 3?

**Reply by Akryum** [reply]
Merging `@vue/composition-api` into Vue 2 core will probably help (with full access to the internals). So mostly beneficial in terms of API support and performance.

**Reply by yyx990803** [reply]
> Do you consider the @vue/composition-api to be feature-complete or are there more apis that need/could be backported in vue 2.7?

It is feature complete, however porting it directly into Vue 2 core will enable less code (direct access to internals removes some hacks needed), better performance, and making writing Composition API code that works in both Vue 2/3 a lot easier since now `import { reactive } from 'vue'` will work the same way across Vue 2/3 (getting rid of the need for adaptor layers like `vue-demi`).

> Will upgrading to 2.7 make it easier to migrate to Vue 3?

Yes, that's another part of the goal for 2.7 (not explicitly mentioned in this RFC). There can be many minor API alignments (e.g. expose the same `createApp` API in Vue 2, support some renamed APIs with deprecation warnings) to ease the migration efforts.

> Will there be an investment in a codemod migrating from the composables written using vue 2 to vue 3?

Technically composables should work the same way in Vue 2/3, since Composition API behavior is expected to be consistent. The reactivity language coverage in Vue 3 is a superset of Vue 2's, so any composables that works in Vue 2 should theoretically just work in Vue 3. (the only thing you probably need to change is where you import the methods from).

**Reply by seanogdev** [reply]
Thanks for that feedback, very happy to hear, especially that you're thinking about not just this step but the migration step onto Vue 3

> Technically composables should work the same way in Vue 2/3, since Composition API behavior is expected to be consistent

The list of [limitations in the current vue 2 composition api](https://github.com/vuejs/composition-api#limitations) requires some less than ideal workarounds that wont be required in Vue 3. 

Will any of these limitations be fixable by access to Vue 2's internals?

**Reply by yyx990803** [reply]
I doubt it, most them are due to ES5 language constraints. Maybe *some*. We'll see as we work on it.

**Comment by AlexandreBonaventure** [comment]
I like that! 
I've always been super curious about the v3 features back-porting to v2. Looking at the rough scope, I'm wondering if the cost of the migration is going to be roughly the same as directly shooting for a complete v3 migration, from an app maintainer standpoint? To me, this backport is going to involve a lot of energy, (slowing down v3 development ?) benefitting only a handful of users (?)(That's an open question, I'm obviously not aware of all the legacy involments nor backed by any sort of metric, but I'd love to see what a survey could reveal here).

**Reply by Akryum** [reply]
We think backporting stuff to Vue 2 will be much easier than supporting IE11 in Vue 3. Worth it for those that still need IE11 support for their Vue apps!

**Reply by AlexandreBonaventure** [reply]
That's fair! So it raises another question:
Is the main motivation for v2 back-porting, the IE11 compat ? And if yes, is it worth it (below 1% and no support from microsoft, IMHO the perk is low)?

**Reply by yyx990803** [reply]
The overall effort shouldn't be huge, since majority of Composition API for Vue 2 is already implemented and validated in `@vue/composition-api`. Outside of that the scope will be have to carefully determined, of course.

This effort is targeting users who have non-trivial codebases that are stuck on Vue 2 due to various reasons (IE11, dependencies, no bandwidth for upgrading). We are pretty sure there are non-trivial amount of existing apps that fall in this group. Our one-time effort could probably benefit many users in such scenarios, saving them the cost of migration altogether.

**Reply by AlexandreBonaventure** [reply]
Glad to hear it. Last question before I go :)
Don't you think this could potentially slow the adoption of Vue3 for both framework usage and the overall ecosystem migration ? Having a viable alternative that is good enough is maybe going to cut the motivation for third-party library maintainers to upgrade for v3 support (which seems to be the main offender according to a totally subjective and non scientific twitter analysis :D)

**Reply by yyx990803** [reply]
The work on 2.7 is planned to happen after an effort to finalize recommending Vue 3 as the default (docs, npm latest tags, etc.). So the goal is for 2.7 to serve as (1) an option for those who absolutely need to support IE and (2) a version that provides enough Vue 3 compatibility to ease the migration efforts for those who want to upgrade to Vue 3.

**Comment by shankhadevpadam** [comment]
When was the vue 2 end of life.

**Reply by leopiccionia** [reply]
For bug fixes, 18 months after the release of Vue 2.7. Some security fixes may be release after that.
https://github.com/vuejs/vue/projects/6#card-16117796

**Comment by egardner** [comment]
I'm a developer at the Wikimedia Foundation, which made the decision to [adopt Vue as our framework of choice](https://phabricator.wikimedia.org/T241180) a little over a year ago. Supporting legacy browsers like IE11 has been a requirement for us since forever basically, although this is [starting to change](https://www.mediawiki.org/wiki/Compatibility/IE11). We have several Vue-powered features in production that are still expected to support IE11 for some time.

**I think this proposal will actually make things easier for us**. If a 2.7 release of Vue is able to back-port some of the main benefits of Vue 3 (Composition API, better typings, etc) while retaining IE11 support, there will be less pressure for us to migrate to Vue 3 right away. I think this would allow us to have a smoother migration path to Vue 3 when we are eventually able to leave IE11 behind us completely.

Thanks for taking the need to support legacy browsers seriously! It was one of the main advantages of Vue in relation to other frameworks as we were evaluating.

**Reply by mbokil** [reply]
I completely agree with this direction. I work for an enterprise medical company and all our apps have to support IE11 still. If I could use the core features of Vue 3 in Vue 2 and later transition to Vue 3 when we move to Chromium Edge in 2 years that would be perfect.

**Comment by zzl110712** [comment]
Will Vue2.7 also support component API and TypeScript in the future?

**Reply by Akryum** [reply]
Yes, that's in the RFC

**Comment by kautilya-pro** [comment]
Drop IE11 support in Vue3, but please provide a fallback for the same (in-built, without a modernizr dependency) just like we do while using `<noscript>` ...

**Reply by leopiccionia** [reply]
The main issue is that Proxy API, widely used in the Vue 3 reactivity implementation, can't be correctly polyfilled in ES5 ([see here](https://github.com/GoogleChrome/proxy-polyfill)). 

So, to support IE, Vue should support two different reactivity implementations:
* One based in `Proxy`, faster and more complete;
* One for IE, with the same reactivity bugs and limitations of Vue 2 (and probably something like `Vue.set` and `Vue.delete`, currently removed in Vue 3).

To Vue core team, supporting IE means a lot of manual code and test writing (it's not just configuring Babel) and the burden of maintaining two different implementations.

To Vue 3 users, supporting IE means you're either actively targetting one subset of Vue reactivity (and using `Vue.set` and `Vue.delete` where otherwise you woudn't need) or writing a lot of duplicated code.

**Reply by kautilya-pro** [reply]
@leopiccionia Thanks for making this easy to understand. But is creating a feature detection out-of-the-box for any unsupported browser hard? I just wanted to have the availability of some fallback code block where we are at least try to tell the user that their browser is unsupported, instead of just showing a broken/partially-working website on all the unsupported websites. I understand there are a lot of other ways to do this on the client end as well as on SSR as well. But getting something out-of-the-box solves a need for some extra dependency library/manual checks.

**Comment by TheJaredWilcurt** [comment]
It would be nice if the summary at the top of this page could be updated to have a nested list of all the Vue 3 features, and their liklihood of being backported. Documenting this at the top would be very helpful.

Otherwise most of these comments will just be asking about specific features. I know that the "Full Rendered Proposal" has a list at the bottom, but no reason to hide the information behind an extra click.

**Comment by niutech** [comment]
I don't think it is a good idea. IE on desktop has [1.7% usage worldwide](https://gs.statcounter.com/browser-market-share/desktop/worldwide/) and [2.2% in North America](https://gs.statcounter.com/browser-market-share/desktop/north-america/). This is millions of users in total, many of which cannot upgrade to a newer web browser due to stupid policies or unsupported configurations. Don't break the web for them! It is too early to drop IE11, let's wait until its global usage on desktop is less than <1%, in time of Vue.js 4.

In comparison, [jQuery](https://jquery.com/browser-support/) and [React](https://reactjs.org/docs/react-dom.html#browser-support) still support IE9+ and [Angular](https://angular.io/guide/browser-support) supports IE11.

**Reply by privatenumber** [reply]
I think you can look at new updates coming to Vue 2.7 as the "IE 11 version of Vue 3" as many Vue 3 features will be back-ported to it (within what's possible with IE 11).

Also read the [perspective](https://github.com/vuejs/rfcs/discussions/296#discussioncomment-563025) from @egardner (Wikimedia dev).

**Reply by niutech** [reply]
@privatenumber Yes, webdevs can still use Vue.js 2.x, but majority of them will hop to 3.x, without thinking about these ~2% users still using IE. So most of the Vue.js webapps will be broken for them, due to negligence. If Vue 3.x supported IE11 out of the box, this would not be a case.

**Reply by darrinmn9** [reply]
The Vue team shouldn't be responsible for the negligence of developers building a website with tools that don't support the browsers their product/business is supposed to support. The docs would be very clear "if you need to support IE 11, use vue 2". That is the entire purpose of Major releases - breaking changes.

**Reply by privatenumber** [reply]
I think in most cases, engineers deliberately opt-out of supporting IE 11 because it adds too much extra work (even to just test support) and would rather join the effort of killing it.

If an engineer unknowingly starts with Vue 3 and realizes they need IE 11 support, hopefully the transition from Vue 3 to Vue 2 isn't too difficult because the gap will be as small as possible.

Also note, the proposal describes that even if Vue 3 supported IE 11, it wouldn't behave the same way because of technical limitations. Therefore, it could lead to a bigger miscommunication problem that conflates two different behaviors under one version. Clearly distinguishing them as v3 and v2.7 will solve this problem.

**Reply by egardner** [reply]
The fact that this proposal includes back-porting some of the big Vue 3 improvements into Vue 2 is key; I also hope that this means Vue 2.7 would continue to be supported for some time going forward. That will allow those of us who do need to support this userbase to continue doing so without missing out on the most important improvements to the framework.

**Reply by mbokil** [reply]
I don't think you can blame developers for IE11. In most cases we have to support it based on business needs and user base. Until everything we have works in a webkit based browser we have no choice to support IE11 for customers. Can't abandon customers business rule #1.

**Reply by kuba108** [reply]
The problem is not in decision to drop support for IE11. Problem is that at the beginning after Vue 3 was released, there was promises that support of IE11 will be in Qx of 2020. I guess it's ok it was delayed, but say: "We are dropping IE11 support" out of blue in beginning of Q2 2021 is just late.

I started to developing new app in Vue 3 and now I will have to rewrite code back to Vue 2.

So once again: It's OK to decide that Vue 3 will not support legacy browsers. However make this decision in time of releasing version, not a half year after.

**Reply by Akryum** [reply]
@kuba108  You would have rewritten code anyway since not all of it would have been compatible with IE11, since Vue 3 would require a different reactivity implementation that works in IE11.

**Reply by kuba108** [reply]
@Akryum If it is so clear now, why it was not clear in September 2020? I thought polyfills will build compatible version with IE and when IE will be gone in about a year from now, I will change polyfill list and voila!

It's very unusual to support and plan development of two major versions at the same time. It reminds me Python and the problem with version 2 and 3. Many of my friends were unable to switch from 2 to 3 over 8 years.

**Reply by Akryum** [reply]
> I thought polyfills will build compatible version with IE and when IE will be gone in about a year from now, I will change polyfill list and voila!

That is a false statement. It was always planned to have a compatibility build with an alternate reactivity system with the same caveats as Vue 2. So not all code made with Vue 3 standard would have worked, that's why we also planned to an hybrid development version to warn for those cases. Some standard 6 years-old JS features are not supported in IE11 and not polyfillable, like Proxy.

> It's very unusual to support and plan development of two major versions at the same time.

We are not though, Vue 2 is already developped. We are continuing supporting it though. The goal is to ease migration of existing projects to Vue 3 and provide an escape hatch for IE11 support.

> It reminds me Python and the problem with version 2 and 3.

???

**Reply by zhantewei2** [reply]
The surport IE is a must in china if you want to make a enterpirse oriented application...saddly..

**Comment by botic** [comment]
I work for a site with ~ 800 million impressions and ~ 15 million unique clients per month in Austria. We see a clear end for IE support and only provide a basic user experience for it. We recently launched a prominent project without initial IE11 support and had not one hateful mail in our inbox.

Since our projects have longer release cycles regarding underlying frameworks and libraries, I don't see a big problem. Existing applications would require a lot of work to update everything to Vue 3, so I'm sure most will stay on Vue 2 till their own EOL. They will keep their existing IE11 support, developer experience is still acceptable.

Regarding the RFC's argument for IE11 for screen reader users: I see the problem, but those users will face a lot of troubles in the coming months, since IE support will fade out for a lot of popular products. If they need IE11 for internal applications, those will not jump on the Vue 3 train either. So for me this should have no effect on a decision, since internal technical debts should stay interal 😃

**Reply by Tofandel** [reply]
IE is usually not a problem for the general public, it's only a problem for legacy entreprise systems, some even still using windows XP that have no other choice than to stay on IE

**Comment by shaal** [comment]
[Drupal core is dropping support for Internet Explorer 11 in Drupal 10](https://www.drupal.org/node/3199540) 

Excerpts from the link above:
**Why are we dropping support for Internet Explorer 11?**
* Microsoft stopped developing Internet Explorer 11 in 2015. For that reason [it is significantly behind modern browsers](https://caniuse.com/?compare=ie+11,firefox+84,chrome+89&compareCats=all%E2%80%9D) that have continued development. Because more and more libraries are adopting the use of those features, there is a significant cost associated with maintaining Internet Explorer 11. For example, the latest major release of CKEditor has dropped support for Internet Explorer 11, which means that Drupal 10 cannot securely support both CKEditor 5 and Internet Explorer 11.
* [Microsoft itself has dropped support for Internet Explorer 11 in many of its services](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666), and the rest will drop support in mid-2021.
Usage of Internet Explorer 11 has decreased significantly. Drupal doesn’t collect data from its end users, so we rely on data provided by other projects such as Wikimedia, WebAIM, and Statcounter. For example, in October 2019, Wikimedia had 4.4% of its traffic using Internet Explorer 11. At the same time this year the number had dropped to 1.4%. We assume that the number of Internet Explorer 11 users will continue decreasing before the release of Drupal 10.
* We consulted several accessibility organisations, and all are either planning to drop support for IE11 soon, or already actively encouraging people to stop supporting and using it.
* Supporting Internet Explorer 11 degrades the experience for everyone. We currently supply all users extra code to make Internet Explorer 11 work. This increases the request size and makes page load time slower for everyone.
* The additional requirements of Internet Explorer 11 demand additional development time that far exceeds the browser's market share. These efforts come at the expense of new features and bug fixes.

You can also read [the discussion that led to that decision](https://www.drupal.org/project/drupal/issues/3155358)

**Comment by dajpes** [comment]
No problem at all for dropping support for IE11, what I'm concerned about tho is the safari explorer, because that browser is gonna be the next internet explorer

**Reply by thedamon** [reply]
I felt this way until recently. Apple scooped up several Firefox devs when Mozilla nearly died last year, and their new focus on privacy makes me pretty optimistic about the browser's future. Though as much as current Safari is looking good, I think the primary issue with web support is now deciding whether to exclude users who are locked to certain older versions of Safari because they are on older mobile hardware many of whom legitimately cannot afford to upgrade (like iPhone 6 which tops out at Safari 12.5.2)

**Comment by ScottAwesome** [comment]
One simple question I have that I didn’t see an answer to in the proposal:

does this mean Vue 2 will have an effective life until 2023? (Edit: I had this date wrong) That’s Microsoft’s own estimated timeline for IE11 to finally be rendered into absolute insignificance that we truly never have to think about it again. This is due to be the year that the last of all support contracts that include OS versions that only shipped with IE11 ends for them (that’s a pretty significant signal as to the life of their EOL platforms)

Of course by that time we will all be complaining about supporting Edge 15 😂

**Comment by ScottAwesome** [comment]
One other factor worth noting: what’s the sum browser total of browsers that don’t support full ES2015 environments? Would this negatively impact users of subregion browsers like Opera Mini, Opera Mobile, QQ Browser, Baidu Browser etc. they have had typically patchy support of the full standard (though I think most of them support the “main” features of ES2015)

**Comment by maximilianschmid** [comment]
Don‘t let you slow down by IE11. Vue2 only IE11 support is great idea. We had heavy IE11 use, 2 months ago we dropped support. No complaints, people just upgrade to Edge or Firefox in our case.

**Comment by cawa-93** [comment]
I have downloaded for you a little raw statistics from several of our sites for the period 2019-2021.

Specifics of sites: 
- Information portal for young mothers.
- Educational encyclopedia for children.
- Forum dedicated to motherhood.

The statistics show:
- Internet Explorer version.
- Country
- Number of visitors from the Internet Explorer.
- Their part of the total number of visitors.


[visitors-browser 2019-2021.zip](https://github.com/vuejs/rfcs/files/6253199/visitors-browser.2019-2020.zip)

**Reply by cawa-93** [reply]
I hope it will be a little useful 😉

**Comment by 2bj** [comment]
for ie11 we have vue2
pls `drop_ie11:true` in vue3

**Comment by ematipico** [comment]
Dropping IE11 is the right move.

Developers should start encouraging their managers that supporting a browsers that has security flaws, won't build any credibility in products they build.

Products and website that support IE11 are not good because they help the users, they are trustworthy because they expose their users to security vulnerabilities.

Vue, drop IE11 for good!

**Comment by designbydc** [comment]
Please drop IE11 support, if not we need also support on C64 with Vue on floppy disk. 
Jokes apart, even Microsoft suggest not to use IE11 anymore and in their Docs they describe since Nov 2020 a new capability that automatically redirects users from IE11 to MS Edge! https://docs.microsoft.com/en-us/microsoft-edge/web-platform/ie-to-microsoft-edge-redirection

**Comment by kris-ellery** [comment]
I strongly support this initiative.

At work, we don't support IE11. However, we do use Vue 2 for our design system and all modern applications. With Vue 2.7 it would be much easier to incrementally migrate our repositories w/out massive breaking changes. Once each repo is 90%+ compatible with Vue 3, take the leap and convert the remaining items.

**Comment by ylynfatt** [comment]
I agree with dropping IE 11 support in Vue 3 and focus on backporting compatible features to Vue 2.7.

**Comment by jshimkoski** [comment]
Microsoft's own 365 apps will [drop support for IE11 in August](https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666) and they have sunset Edge Legacy (non-chrome). IE11 has been dead for awhile and is a [compatibility solution](https://techcommunity.microsoft.com/t5/windows-it-pro-blog/the-perils-of-using-internet-explorer-as-your-default-browser/ba-p/331732).

Do not dirty up the codebase to support this dinosaur.

**Comment by SirMishaa** [comment]
IE11 support requires too much work, does not bring any additional gain (as we see, large companies currently drop IE) and, due to polyfills increase the weight of VueJS and sometimes reduce performance.

I think that in 2021 is the right time to start the discontinuation of support for Internet Explorer 11. Developers who need to maintain IE11 support will continue on version 2.7 and still keep support.

It's great that Vue is thinking about the users who are still using IE11 and still offering support for Vue 2.

**Comment by jods4** [comment]
I expected this outcome since I jumped into Vue 3 betas. It is the reasonable thing to do, for the reasons stated in your RFC. 👍 

Front-end frameworks rewrite themselves every 5 years or so. It's this time for Vue and you must write Vue 3 for the next 5 years, not for browsers from 5 years ago. When IE support is needed, it's perfectly ok to continue using Vue 2 for the next 2-3 years. Hopefully IE support won't be needed anymore by then.

I'm supportive of this choice despite the fact that I am actually maintaining 2 projects that still support IE 11. You can't ask to have the latest innovations when you're using something that's 8 years old.

BTW if you want anecdotal data: I work for a large software company in central Europe. We make hundreds of projects each year for middle to large enterprises in all sectors. These are your typical "enterprise" projects where IE was the only browser available a few years back. Nobody starts new IE projects today and all clients are migrating to newer browsers at a really fast pace for the past ~2 years.
For instance the two 6-years-old IE projects that I'm personally responsible for are already compatible with Chrome and we'll probably drop IE altogether for 2022.

**Comment by sirlancelot** [comment]
I would like to see an opt-in way to support the new `v-model` mechanics for custom components. Specifically, having the prop/event become `modelValue` and `@update:modelValue` (and similarly with `v-model:custom`). To me, this seems like it would be the biggest pain point for upgrading my codebases from 2 -> 3. Being able to piecemeal upgrade each component on my own time with a transitional 2.x build would do wonders in lowering the barrier of entry for a 2 -> 3 transition.

**Reply by sirlancelot** [reply]
On second thought, maybe Vue 2's `model` component option can already aid in this:

```js
export default {
  model: { prop: "modelValue", event: "update:modelValue" },
  props: ["modelValue"],
}
```

That should do it for bare `v-model`, but I still have _maybe_ half a dozen areas where `:custom.sync` is used. It would be nice to be able to convert those to `v-model:custom`.

**Comment by eskimm** [comment]
There is no benefit for Vue 3 to support it. Vue 2 will do it.

**Comment by pedro-gilmora** [comment]
Why should we bring support for something that is a constant pain ... Every new native HTML/CSS/JS features not supported and not intended to ship on IE will be headache to move to IE... Pollyfills might help... But always comes with performance, undesired larger implementation and other lacks to the Vue development. It would be nice to have it as a Vue/CLI option to not install too much pollyfills in case IE support  will be a fact

**Comment by yyx990803** [comment]
It seems the community is overwhelmingly in support of dropping IE11 support.

This RFC is now in final comments stage. An RFC in final comments stage means that:

The core team has reviewed the feedback and reached consensus about the general direction of the RFC and believe that this RFC is a worthwhile addition to the framework.
Final comments stage does not mean the RFC's design details are final - we may still tweak the details as we implement it and discover new technical insights or constraints. It may even be further adjusted based on user feedback after it lands in an alpha/beta release.
If no major objections with solid supporting arguments have been presented after a week, the RFC will be merged and become an active RFC.

**Comment by BlakeB415** [comment]


**Reply by ghost** [reply]
I don't agree. Even Android 4.4 support Chrome 81. Users using Android below 4.4 are less than 0.5%

**Reply by jods4** [reply]
> The main issue I see is that some plugin developers are completely dropping support for Vue 2 in newer versions of their plugins. So even if we have Vue 2 for ES5 support, newer plugins may not support it.

The thing is, even if you had Vue 3 with an ES property-based reactivity module in place of the proxy one, those shiny Vue 3 plugins would most likely still not work for you.
 
 Those plugin devs that dropped Vue 2 won't restrain themselves to use IE-11 compatible reactivity API only.

The fact that proxies can't be emulated in older browser is the deal breaker. IE11-compatible-Vue3, if it existed, would not be the same framework as Vue 3 and most libraries/plugins written for Vue 3 wouldn't work in IE 11.

> Though by rough estimates, that is around 125,000,000 android devices

Which is still 0.5%, even if the absolute number looks bigger.

**Reply by BlakeB415** [reply]
> > The main issue I see is that some plugin developers are completely dropping support for Vue 2 in newer versions of their plugins. So even if we have Vue 2 for ES5 support, newer plugins may not support it.
> 
> The thing is, even if you had Vue 3 with an ES property-based reactivity module in place of the proxy one, those shiny Vue 3 plugins would most likely still not work for you.
> 
> Those plugin devs that dropped Vue 2 won't restrain themselves to use IE-11 compatible reactivity API only.
> 
> The fact that proxies can't be emulated in older browser is the deal breaker. IE11-compatible-Vue3, if it existed, would not be the same framework as Vue 3 and most libraries/plugins written for Vue 3 wouldn't work in IE 11.
> 
> > Though by rough estimates, that is around 125,000,000 android devices
> 
> Which is still 0.5%, even if the absolute number looks bigger.

That's understandable. For most websites, supporting IE is not a necessity and due to the technical limitations I can definitely see why dropping IE will probably be the best course of action in this scenario. Unfortunately, this means I'll have to look elsewhere for another framework/library that will fit our needs. Thanks for taking your time to reply.

**Reply by privatenumber** [reply]
Confused why you need to move to a different framework? They said they're going to maintain Vue 2.7 which has IE 11 support.

If ecosystem size/support is your primary issue, your options for another framework is quite limited. I can't imagine the other major frameworks supporting IE11 in the next few years either.

The Vue ecosystem primarily supports Vue 2 right now, but by the time it's predominantly Vue 3 in a year or two, you might not have the IE requirement. Even if you do, I think it will take less work to fork the necessary plugins and backport patches/features yourself.

**Reply by cjblomqvist** [reply]
> > I don't agree. Even Android 4.4 support Chrome 81. Users using Android below 4.4 are less than 0.5%
> 
> That’s true. Though by rough estimates, that is around 125,000,000 android devices based on the active number of android devices (2.5 billion) recorded in 2019 (which is the most up-to-date statistic for that) as well as the percentage you gave. Which isn’t much in the grand scheme of things but still somewhat significant. For IOS 9, its current market share as of 2021 is around 1% of the IOS user base. Which comes out at around 150,000,000 devices based on the 1.5 billion active IOS devices stat.

Your numbers are off by a magnitude. 1% of 1.5 billion = 1500 million / 100 = 15 million. Quite the difference from 150 million....

**Reply by BlakeB415** [reply]
> > > I don't agree. Even Android 4.4 support Chrome 81. Users using Android below 4.4 are less than 0.5%
> > 
> > 
> > That’s true. Though by rough estimates, that is around 125,000,000 android devices based on the active number of android devices (2.5 billion) recorded in 2019 (which is the most up-to-date statistic for that) as well as the percentage you gave. Which isn’t much in the grand scheme of things but still somewhat significant. For IOS 9, its current market share as of 2021 is around 1% of the IOS user base. Which comes out at around 150,000,000 devices based on the 1.5 billion active IOS devices stat.
> 
> Your numbers are off by a magnitude. 1% of 1.5 billion = 1500 million / 100 = 15 million. Quite the difference from 150 million....

You’re absolutely right! My bad. Accidentally put 0.1 instead of 0.01 into the calculator in a rush.

**Comment by martinszeltins** [comment]
I totally agree with this proposal. I think IE is dragging down in innovation and just makes the developer's life just plain frustrating. Can't remember the last project where I had to support IE 11.

**Comment by cjblomqvist** [comment]
I believe this is reasonable course of action. What I'm more afraid of is the how well Vue 2 will be maintained, and possibly for how long. I understand it's a big burden to maintain (and further develop) all the libraries you do, and you do it (sort of) for free, but the activity in the Vue 2 repo has been really slow since Vue 3 was released, as can easily be seen by the number of commits vs issue and PR trends. Issues such as https://github.com/vuejs/vue/issues/11718 should have been fixed a long time ago. (I know vite and all that, but the fact is that webpack is the main bundler atm and vite is the new unknown kid on the block) 

Perhaps a promise to _actively_ maintain Vue 2 until IE11 has less than 0.X% of global usage would help? And show some proper effort into Vue 2 until that day happens.

**Comment by lllllxt** [comment]
我也同意放弃IE，IE简直拖后腿。但还是希望以后有人可以出个vue3降级vue2的指南，我曾遇到一个项目，开发初期和甲方再三确认过不需要兼容IE，但在项目二期的时候他们又反悔了，鉴于种种原因我们没有拒绝的权力。

**Reply by jpg1024** [reply]
赞同,Vue 3 不支持IE,无可厚非.但希望Vue 2的维护时间能长一些.

**Reply by ChasLui** [reply]
> 赞同,Vue 3 不支持IE,无可厚非.但希望Vue 2的维护时间能长一些.

vue2 和 vue3 应该作为长期维护版本并行

**Reply by Darcrandex** [reply]
科技强国的底气[doge]

**Reply by lishihong** [reply]
后面怎么解决的

**Comment by tnikolai2** [comment]
ios9 will also not be supported?

**Reply by BlakeB415** [reply]
I’d assume so considering Proxy objects are not supported on that version.

**Comment by Aaron-Pool** [comment]
It's been 10 days since the RFC reached final comments status. Any major holdup or particular reason it hasn't been merged?

**Comment by jpg1024** [comment]
Unfortunately, my organization will continue to use IE11 for now and in the next few years. What I am more concerned about is, will Vue 2x continue to add new features and bug fixes? Or will it stop adding new features after a period of time and only keep bugs Fix it? Thanks!

-

很遗憾,我所在的机构目前以及今后几年都将继续使用IE11.我比较关心的是,Vue 2x会一直持续添加新特性和Bug修复吗?还是会在一段时间后停止添加新特性仅保留Bug修复?谢谢!

**Comment by kristjanjansen** [comment]
FYI "RFC: Internet Explorer 11 support deprecation and removal" https://github.com/angular/angular/issues/41840

**Comment by simon04** [comment]
The [vue-next readme](https://github.com/vuejs/vue-next#changes-from-vue-2) should be updated accordingly ("Note: IE11 support is still pending.")

**Comment by yyx990803** [comment]
Considering there has been no major objection, merging.

Regarding 2.x maintenance: we will also pick off some long standing 2.x bugs while working on 2.7, and 2.7 will be maintained for at least 18 months.

**Comment by yyx990803** [comment]
This RFC is now finalized.

https://github.com/vuejs/rfcs/blob/master/active-rfcs/0038-vue3-ie11-support.md

**Comment by Haroenv** [comment]
> The complexity would still be somewhat acceptable if it can be fully contained within Vue itself. However, after discussing with community members, we realized the co-existence of two reactivity implementations inevitably leaks to library authors as well.
> 
> By supporting IE11 in Vue 3, library authors essenitally need to make that call as well. Library authors will have to account for what build of Vue 3 their library is running with (on top of potentially also supporting Vue 2) - and if they decide to support IE11, they have to author their library with all the ES5 reactivity caveats in mind.

Due to this big change, I assume a lot of organisations will be stuck on Vue 2 for a while to come. This means that we need to make libraries that support both vue 2 and vue 3 or maintain two branches for libraries. Both of those options aren't super clear to me at the moment. Are there any examples of libraries supporting both versions?

I totally understand why IE11 isn't supported, but I fear a python 2/3 rift is easy to happen. The previous major version was in 2016 when Vue was much smaller, and there weren't as many libraries yet.

**Reply by ChrisShank** [reply]
One option to support both is vue-demi. https://github.com/vueuse/vue-demi

**Reply by Haroenv** [reply]
Vue demi definitely helps! What I'm struggling with is deciding to use both babel transforms and have separate entry points for vue 2 & 3 (and what to do with files that use createElement instead of templates). I know this isn't really the right place to ask, but I'm wondering if there's any libraries that support both I can look at @ChrisShank? Thanks!

**Reply by ChrisShank** [reply]
You should check out how `vue-use`handles it. (I believe they use vue-demi under the hood) 

https://github.com/vueuse/vueuse

**Comment by chenyulun** [comment]
[proxy attribute support](https://caniuse.com/?search=proxy)
It says in the property list：
ios>10
chrome>49

This has nothing to do with IE，We need to support iOS 9 on our mobile devices.

Too bad. Hope iOS9 and lower version Android devices will be phased out in a few years.

**Reply by Akryum** [reply]
> This has nothing to do with IE

Why?

**Reply by chenyulun** [reply]
I want to say that the reason IE is not supported is because VUE3 uses ES6 features，Our team does more business on mobile terminal，Therefore, for our team, we will pay more attention to the lower version of mobile phones，
That means we don't have to worry about supporting IE

**Reply by Akryum** [reply]
I'd say that the reasoning is the same between having to support IE11 and having to support other old browsers. The same [cost](https://github.com/vuejs/rfcs/blob/ie11/active-rfcs/0038-vue3-ie11-support.md#the-cost-of-supporting-ie11-in-vue-3) applies.

**Reply by chenyulun** [reply]
Yes, I agree！It's okay to be radical

**Comment by shaal** [comment]
Microsoft announced today the official EOL of IE11 - June 22, 2022.

https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge/

**Comment by ruffin--** [comment]
> [Please read the full RFC.](https://github.com/vuejs/rfcs/discussions/296#discussioncomment-562796)

Fwiw, the link to the RFC has [moved](https://github.com/vuejs/rfcs/blob/c787b105f7561315b7c14f44399b96cc1654672e/active-rfcs/0038-vue3-ie11-support.md). (Well, it's no longer active, but this is the repo's archived copy.) Good read about IE11 considerations in general; thanks.

Here are a few highlights from the RFC for TL;DC[lick]:

> Vue 2's reactivity system is based on ES5 getter/setters. Vue 3 leverages ES2015 Proxies for a more performant and complete reactivity system, which cannot be polyfilled in IE11.
> ...
> The complexity would still be somewhat acceptable if it can be fully contained within Vue itself. However, after discussing with community members, we realized the co-existence of two reactivity implementations inevitably leaks to library authors as well.
> ...
> Our recommendation is to use Vue 2 if you need IE11 support. Instead of incurring significant technical debt for Vue 3 and future versions of Vue, **we believe it would be more meaningful to redirect the effort towards backporting compatible features to Vue 2** in the 2.7 release, and ensure a closer develpment experience across the two major versions. [emph mine]
